pkg/cache/cache_test.go

```diff
@@ -66,188 +116,165 @@ var _ = Describe("Informer Cache", func() {
 				}
 			}
 			Expect(hasKubeService).To(BeTrue())
+			close(done)
+		})
+
+		It("should be able to get objects that haven't been watched previously", func(done Done) {
+			By("getting the Kubernetes service")
+			svc := &kcorev1.Service{}
+			svcKey := client.ObjectKey{Namespace: "default", Name: "kubernetes"}
+			Expect(informerCache.Get(context.Background(), svcKey, svc)).NotTo(HaveOccurred())
+
+			By("verifying that the returned service looks reasonable")
+			Expect(svc.Name).To(Equal("kubernetes"))
+			Expect(svc.Namespace).To(Equal("default"))
+			close(done)
+		})
+
+		It("should support filtering by labels", func(done Done) {
+			By("listing pods with a particular label")
+			// NB: each pod has a "test-label" equal to the pod name
+			out := kcorev1.PodList{}
+			Expect(informerCache.List(context.TODO(), client.InNamespace("test-namespace-2").
+				MatchingLabels(map[string]string{"test-label": "test-pod-2"}), &out)).NotTo(HaveOccurred())
+
+			By("verifying the returned pods have the correct label")
+			Expect(out.Items).NotTo(BeEmpty())
+			Expect(len(out.Items)).To(Equal(1))
+			actual := out.Items[0]
+			Expect(actual.Labels["test-label"]).To(Equal("test-pod-2"))
+			close(done)
+		})
+
+		It("should be able to list objects by namespace", func(done Done) {
+			By("listing pods in test-namespace-1")
+			listObj := &kcorev1.PodList{}
+			Expect(informerCache.List(context.Background(),
+				client.InNamespace("test-namespace-1"),
+				listObj)).NotTo(HaveOccurred())
+
+			By("verifying that the returned pods are in test-namespace-1")
+			Expect(listObj.Items).NotTo(BeEmpty())
+			Expect(len(listObj.Items)).To(Equal(1))
+			actual := listObj.Items[0]
+			Expect(actual.Namespace).To(Equal("test-namespace-1"))
+			close(done)
+		})
+
+		It("should deep copy the object unless told otherwise", func(done Done) {
+			By("retrieving a specific pod from the cache")
+			out := &kcorev1.Pod{}
+			podKey := client.ObjectKey{Name: "test-pod-2", Namespace: "test-namespace-2"}
+			Expect(informerCache.Get(context.Background(), podKey, out)).NotTo(HaveOccurred())
+
+			By("verifying the retrieved pod is equal to a known pod")
+			Expect(out).To(Equal(knownPod2))
+
+			By("altering a field in the retrieved pod")
+			*out.Spec.ActiveDeadlineSeconds = 4
+
+			By("verifying the pods are no longer equal")
+			Expect(out).NotTo(Equal(knownPod2))
+			close(done)
+		})
+
+		It("should error out for missing objects", func(done Done) {
+			By("getting a service that does not exists")
+			svc := &kcorev1.Service{}
+			svcKey := client.ObjectKey{Namespace: "unknown", Name: "unknown"}
+
+			By("verifying that an error is returned")
+			Expect(informerCache.Get(context.Background(), svcKey, svc)).To(HaveOccurred())
```

Comment:
check that the error is a not found error (kube errors IsNotFound).  we actually want to guarantee that contract.